### ISSUE TITLE
It is not possible to create aliases of a function within another function

### ISSUE THREAD

**Issue by ErickTucto** [xonsh/xonsh#4038]
I try to create aliases using function, but xonsh used in python3.9 doesn't work, while in python3.8 it works, I share screenshots

**screenshot (python 3.8)**

![python38](https://user-images.githubusercontent.com/22831024/103179312-b96a3780-4858-11eb-8943-300bd98d838c.png)

**screenshot (python 3.9)**

![arch](https://user-images.githubusercontent.com/22831024/103179328-d6066f80-4858-11eb-912e-2c55e2ed04c9.png)

The error causes it
```python
aliases[function.__name__] = function
```

## xonfig

<details>
python3.8
+------------------+-----------------+
| xonsh            | 0.9.24          |
| Python           | 3.8.5           |
| PLY              | 3.11            |
| have readline    | True            |
| prompt toolkit   | None            |
| shell type       | prompt_toolkit  |
| pygments         | 2.7.3           |
| on posix         | True            |
| on linux         | True            |
| distro           | linuxmint       |
| on darwin        | False           |
| on windows       | False           |
| on cygwin        | False           |
| on msys2         | False           |
| is superuser     | False           |
| default encoding | utf-8           |
| xonsh encoding   | utf-8           |
| encoding errors  | surrogateescape |
| on jupyter       | False           |
| jupyter kernel   | None            |
+------------------+-----------------+

python3.9
+------------------+-----------------+
| xonsh            | 0.9.24          |
| Python           | 3.9.1           |
| PLY              | 3.11            |
| have readline    | True            |
| prompt toolkit   | None            |
| shell type       | prompt_toolkit  |
| pygments         | 2.7.3           |
| on posix         | True            |
| on linux         | True            |
| distro           | arch            |
| on darwin        | False           |
| on windows       | False           |
| on cygwin        | False           |
| on msys2         | False           |
| is superuser     | False           |
| default encoding | utf-8           |
| xonsh encoding   | utf-8           |
| encoding errors  | surrogateescape |
| on jupyter       | False           |
| jupyter kernel   | None            |
+------------------+-----------------+

</details>

## Expected Behavior

Create an alias to a function using the function name

## Current Behavior

It causes an error even though the python interpreter works. I share capture.

![python-inter](https://user-images.githubusercontent.com/22831024/103179416-e2d79300-4859-11eb-88cc-50d6c4d635ac.png)

### Traceback (if applicable)

<details>

```
Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/xonsh/ptk_shell/shell.py", line 238, in \_push
    code = self.execer.compile(src, mode="single", glbs=self.ctx, locs=None)
  File "/usr/lib/python3.9/site-packages/xonsh/**amalgam**.py", line 23992, in compile
    tree = self.parse(input, ctx, mode=mode, filename=filename, transform=transform)
  File "/usr/lib/python3.9/site-packages/xonsh/**amalgam**.py", line 23952, in parse
    tree, input = self.\_parse_ctx_free(input, mode=mode, filename=filename)
  File "/usr/lib/python3.9/site-packages/xonsh/**amalgam**.py", line 24076, in \_parse_ctx_free
    tree = self.parser.parse(
  File "/usr/lib/python3.9/site-packages/xonsh/parsers/base.py", line 497, in parse
    tree = self.parser.parse(input=s, lexer=self.lexer, debug=debug_level)
  File "/usr/lib/python3.9/site-packages/xonsh/ply/ply/yacc.py", line 335, in parse
    return self.parseopt_notrack(input, lexer, debug, tracking, tokenfunc)
  File "/usr/lib/python3.9/site-packages/xonsh/ply/ply/yacc.py", line 1122, in parseopt_notrack
    p.callable(pslice)
  File "/usr/lib/python3.9/site-packages/xonsh/parsers/base.py", line 2221, in p_atom_expr
    p[0] = self.apply_trailers(p[1], p[2])
  File "/usr/lib/python3.9/site-packages/xonsh/parsers/base.py", line 2215, in apply_trailers
    assert False
AssertionError
```

</details>

## Steps to Reproduce

![arch](https://user-images.githubusercontent.com/22831024/103179436-0a2e6000-485a-11eb-93b5-ec5512dfe586.png)

## For community

⬇️ **Please click the 👍 reaction instead of leaving a `+1` or 👍 comment**

**Comment by gforsyth** [comment]
Thanks for reporting @ErickTucto -- we don't have explicit 3.9 support yet and there were several changes made to the AST.  I'll add this to the list.

**Comment by anki-code** [comment]
Fixed in master by #4101.


Tested in docker:
```python
docker run --rm -it python:3.9-slim /bin/bash -c "apt update && apt install -y git && pip install -U git+https://github.com/xonsh/xonsh.git && xonsh"
```
